### ISSUE TITLE
`not-an-iterable` and `unsubscriptable-object` false positives with @overload method

### ISSUE THREAD

**Issue by Jonas1312** [pylint-dev/pylint#8634]
### Bug description

I'm having two errors `not-an-iterable` and `unsubscriptable-object` with the code below:

```python
from typing import Generator, Iterable, List, Union, overload

from typing_extensions import Literal, assert_type


class Format1:
    def __init__(self, *args, **kwargs) -> None:
        self.data = (args, kwargs)


class Format2:
    def __init__(self, *args, **kwargs) -> None:
        self.data = (args, kwargs)


class MyClass:
    def __init__(self) -> None:
        pass

    @overload
    def my_method(
        self,
        output_format: Literal["format_1"] = "format_1",
        *,
        as_generator: Literal[False] = False
    ) -> List[Format1]:
        ...

    @overload
    def my_method(
        self,
        output_format: Literal["format_2"] = "format_2",
        *,
        as_generator: Literal[False] = False
    ) -> List[Format2]:
        ...

    @overload
    def my_method(
        self, output_format: Literal["format_1"] = "format_1", *, as_generator: Literal[True] = True
    ) -> Generator[Format1, None, None]:
        ...

    @overload
    def my_method(
        self, output_format: Literal["format_2"] = "format_2", *, as_generator: Literal[True] = True
    ) -> Generator[Format2, None, None]:
        ...

    def my_method(
        self,
        output_format: Literal["format_1", "format_2"] = "format_1",
        *,
        as_generator: bool = False
    ) -> Iterable[Union[Format1, Format2]]:
        some_data = [1, 2, 3]

        format_class = Format1 if output_format == "format_1" else Format2
        gen = (format_class(i) for i in some_data)
        if as_generator:
            return gen
        return list(gen)


def test_1() -> None:
    assert_type(MyClass().my_method(), List[Format1])
    assert_type(MyClass().my_method(as_generator=True), Generator[Format1, None, None])
    assert_type(MyClass().my_method(output_format="format_2"), List[Format2])
    assert_type(
        MyClass().my_method(output_format="format_2", as_generator=True),
        Generator[Format2, None, None],
    )


def test_2() -> None:
    my_obj = MyClass()
    ret = my_obj.my_method()
    print(ret)

    assert_type(ret, List[Format1])

    assert isinstance(ret, List)
    assert all(isinstance(x, Format1) for x in ret)  # pylint not-an-iterable error here
    assert ret[0].data == ((1,), {})  # pylint unsubscriptable-object error here


if __name__ == "__main__":
    test_2()

```

pyright does not throw any errors with the code written in `test_1`. So pyright is able to figure out the proper return types thanks to the different `@overload`.

But pylint raises some errors in `test_2`, even though the code runs well.

I am not sure if it's a pylint bug, or if I didn't `@overload` the method `my_method` properly.

Thanks

### Command used

```shell
pylint mre.py
```

### Pylint output

```shell
mre.py:89:47: E1133: Non-iterable value ret is used in an iterating context (not-an-iterable)
mre.py:90:11: E1136: Value 'ret' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

pylint should not raise those errors since pyright can identify the return types properly, and the test_2 passes

### Pylint version

```shell
pylint 2.17.3
astroid 2.15.4
Python 3.7.15 (default, Nov 24 2022, 12:02:37)
```

### OS / Environment

macos

### Additional dependencies

pyright==1.1.293

**Comment by mbyrnepr2** [comment]
Thanks for the report @Jonas1312! I think this is a duplicate of https://github.com/pylint-dev/astroid/issues/1015.

**Comment by mbyrnepr2** [comment]
There is also https://github.com/pylint-dev/pylint/issues/5712.

**Comment by Jonas1312** [comment]
My bad! I didn't find the already existing issues... 🙏
